### PR TITLE
feat: assign channels to servers

### PR DESCRIPTION
**Pull request by GeraldLoeffler**
---
title: "Assign channels to servers through their names"
---

Add `servers` field to channel items, so that channels can be assigned to a sub-set of servers using the names of those servers. (A server's name is the string key used for defining that server in the top-level `servers` object.)

Solves the issues discussed in the related issues:

---
**Related issue(s):**

#244 
#255 
#475 
---

### CODE REVIEW

**Comment by smoya** [comment]
I like the approach of using **labels** here. In fact, It made me think about Kubernetes labels and selectors. 
Perhaps, we could move one step further and add the same `metadata` concept for servers so labels can be a map of string => string as well. In this way we could also tag servers with labels such as `env: production` or similar.

**Comment by fmvilas** [comment]
@smoya I think going the same path as k8s is a bit of a premature optimization for now IMHO. AFAIK, nobody has asked for a more sophisticated mechanism, and adding it will introduce potentially unnecessary complexity to the spec.

**Comment by derberg** [comment]
@GeraldLoeffler could you clarify more about why it is Server and Channel relation and not Server and Message relation? There is a use case that one channel can have multiple messages on a given operation (`messages.oneOf`). You can have 2 servers with the same channel, but some messages are available only on 1 server. What would be the possible problem if we have `serverLabel` on the message and not the channel?

Another suggestion from my side would be a bit more modification on the example. Now `labels` kind of duplicate information like server key and protocol. This may cause confusion.

Another thinking exercise is: at first sight `labels` seem to be nifty solution but is it really necessary here, do we have a use case for so many servers that they need some kind of label? wouldn't it be better -> the channel/message just have a `server` object that is a list of servers? we would leverage `$ref` usage here of course? I think it would be more aligned with the spec 🤔 like that users could reuse existing solutions like `$ref` and `traits` to optimize AsyncAPI document and keep it DRY, instead of the `label` system.

Sorry for all these challenges 😅 

Huge additional request, please explicitly mention this PR as a comment in all the linked issues so people that are interested the most are notified.

**Comment by GeraldLoeffler** [comment]
> @GeraldLoeffler could you clarify more about why it is Server and Channel relation and not Server and Message relation? There is a use case that one channel can have multiple messages on a given operation (`messages.oneOf`). You can have 2 servers with the same channel, but some messages are available only on 1 server. What would be the possible problem if we have `serverLabel` on the message and not the channel?

it was simply a question of modelling the most obvious association, the one between server and channel:
- a queue/topic (= channel) is inherently (often statically) bound to a message broker (= server)
- a URL path (= channel) is inherently (typically statically) bound to a HTTP server (= server)
So far in the spec it was assumed that in the scope of a given AsyncAPI document, all channels are available on all servers. I'm trying to relax that requirement.

You are saying that the same channel may exist on several servers, but some messages may only be sent/received via that channel to/from a subset of those servers. True, that can happen too, but i didn't have the need to model that and so it's not in my proposal. If we wanted to include it we would need an optional `serverLabels` field on the message (in addition to the one on the channel).

**Comment by GeraldLoeffler** [comment]
> Another suggestion from my side would be a bit more modification on the example. Now `labels` kind of duplicate information like server key and protocol. This may cause confusion.

agreed, the examples could be better: happy to refine them once we have agreement on the approach.

**Comment by GeraldLoeffler** [comment]
> Another thinking exercise is: at first sight `labels` seem to be nifty solution but is it really necessary here, do we have a use case for so many servers that they need some kind of label? wouldn't it be better -> the channel/message just have a `server` object that is a list of servers? we would leverage `$ref` usage here of course? I think it would be more aligned with the spec 🤔 like that users could reuse existing solutions like `$ref` and `traits` to optimize AsyncAPI document and keep it DRY, instead of the `label` system.

your proposal is to bind channels to servers by explicitly listing all servers to which a channel applies. This is certainly possible, but cumbersome in the typical case of having many environments. For example, for a protocol translator from `mqtt` to `jms`:
```
servers:
 mqtt-dev:
    url: ...
  mqtt-test:
    url: ...
  mqtt-prod:
    url: ...
  jms-dev:
    url: ...
  jms-test:
    url: ...
  jms-prod:
    url: ...

from-channel-1:
  servers:
    - $ref "...mqtt-dev"
    - $ref "...mqtt-test"
    - $ref "...mqtt-prod"
  ...
from-channel-2:
  servers:
    - $ref "...mqtt-dev"
    - $ref "...mqtt-test"
    - $ref "...mqtt-prod"
  ...
to-channel-1:
  servers:
    - $ref "...jms-dev"
    - $ref "...jms-test"
    - $ref "...jms-prod"
  ...
to-channel-2:
  servers:
    - $ref "...jms-dev"
    - $ref "...jms-test"
    - $ref "...jms-prod"
  ...
```
versus
```
servers:
 mqtt-dev:
    url: ...
    label: ["from"]
  mqtt-test:
    url: ...
    label: ["from"]
  mqtt-prod:
    url: ...
    label: ["from"]
  jms-dev:
    url: ...
    label: ["to"]
  jms-test:
    url: ...
    label: ["to"]
  jms-prod:
    url: ...
    label: ["to"]

from-channel-1:
  serverLabels: ["from"]
  ...
from-channel-2:
  serverLabels: ["from"]
  ...
to-channel-1:
  serverLabels: ["to"]
  ...
to-channel-2:
  serverLabels: ["to"]
  ...
```
I think this example shows the tradeoff between the two approaches.

**Comment by smoya** [comment]
> @GeraldLoeffler could you clarify more about why it is Server and Channel relation and not Server and Message relation? There is a use case that one channel can have multiple messages on a given operation (`messages.oneOf`). You can have 2 servers with the same channel, but some messages are available only on 1 server. What would be the possible problem if we have `serverLabel` on the message and not the channel?

Would it make sense to add, instead, an extra link between Servers and Operations? This is something we have considered when designing the new Intent-driven API for the Parser(s). See https://github.com/asyncapi/parser-api/blob/master/docs/v1.md#operation.

However, having several resources linking to servers can introduce issues such as mutual exclusion. What if a channel specifies its relation to some servers but their operations are linked to others?

**Comment by GeraldLoeffler** [comment]
Calling all interested parties to please **support one of the following proposals** on this topic:

1. associate channels to servers via simple string labels on servers ([original proposal](https://github.com/asyncapi/spec/pull/531/files) by @GeraldLoeffler, also shown in the 2nd part [here](https://github.com/asyncapi/spec/pull/531#issuecomment-828252634))
2. ditto, but extend the labelling scheme to be name-value pairs exactly as in Kubernetes ([proposal](https://github.com/asyncapi/spec/pull/531#issuecomment-827782537) by @smoya)
3. associate channels to servers by referring to the servers directly, without any labels (proposal by @derberg, shown in the 1st part [here](https://github.com/asyncapi/spec/pull/531#issuecomment-828252634))

In addition, there were proposal to go beyond the association of channels to servers:

4. also associate operations to servers ([proposal](https://github.com/asyncapi/spec/pull/531#issuecomment-828293409) by @smoya)
5. also associate messages to servers ([proposal](https://github.com/asyncapi/spec/pull/531#issuecomment-828238520) by @derberg)

Proposals 4. and 5. are additional associations. So the mechanism (string labels, key-value-pair labels, or direct reference) to link to servers in 4. and 5. should match whatever is chosen to link channels to servers. I therefore propose to cover 4.-5. by a separate PR and focus here on 1.-3. .

Please vote on 1., 2., or 3., and also state your support for 4. and/or 5. (@smoya, @derberg, @fmvilas).

**Comment by tyazid** [comment]
> > @GeraldLoeffler could you clarify more about why it is Server and Channel relation and not Server and Message relation? There is a use case that one channel can have multiple messages on a given operation (`messages.oneOf`). You can have 2 servers with the same channel, but some messages are available only on 1 server. What would be the possible problem if we have `serverLabel` on the message and not the channel?
> 
> Would it make sense to add, instead, an extra link between Servers and Operations? This is something we have considered when designing the new Intent-driven API for the Parser(s). See https://github.com/asyncapi/parser-api/blob/master/docs/v1.md#operation.
> 
> However, having several resources linking to servers can introduce issues such as mutual exclusion. What if a channel specifies its relation to some servers but their operations are linked to others?

Rather consider the association of an operation with a server as optional because it is not the general case. Indeed, an operation should (and it is) be associated with a message which can be exchanged with one or several servers... in my opinion.

**Comment by DavidBiesack** [comment]
See also a similar request for OpenAPI to define reusable sets of parameters (or parameter sets). There may be some ideas presented there that you can use to refine this feature proposal.  I.e. labels are juts one way to define a set of servers (all servers sharing a label constitute a server set), and you want to be able to apply or associate a channel to one or more server sets.

**Comment by derberg** [comment]
@GeraldLoeffler I think it is hard to perform any voting if there are still some unknowns. 

> agreed, the examples could be better: happy to refine them once we have agreement on the approach.

I think good example is necessary for others to follow proposal

> your proposal is to bind channels to servers by explicitly listing all servers to which a channel applies. This is certainly possible, but cumbersome in the typical case of having many environments.

The advantage is that you have validation out of the box, parsing fails if you want to parse outdated reference. With labeling, one can change a label on a channel to a name that is not on the server list. Of course this could also be validated but therefore needs to be specified in the spec so parser know it is mandatory to implement.

---

Can you provide a full example of AsyncAPI file with thes new properties. I'd like to have a closer look on a real use case and see how it works for websocket. We have over representation of people supporting concepts with broker in the middle and I just want to explore how this would work for websocket, if this is of any use actually.

> also associate messages to servers (proposal by @derberg)

Let's drop it. I see this in websockets only and for the server code generation where in fact in my opinion this should be really solved by having 2 separate spec documents, with external common parts

**Comment by GeraldLoeffler** [comment]
@derberg @smoya , @fmvilas : going the path of least resistence here and adopting @derberg's proposal, i.e., associating channels to servers by referring to the servers directly, without any labels. This works and is the simplest possible solution.

Have also added a (i think) convincing example:

```
servers:
  webSocketsDev:
    $ref: "#/components/servers/wsDev"
  webSocketsProd:
    $ref: "#/components/servers/wsProd"
  amqpProd:
    $ref: "#/components/servers/rabbitmqBrokerInProd"
  amqpStage:
    $ref: "#/components/servers/rabbitmqBrokerInStaging"

webUICommandsQueue:
  description: This application publishes WebUICommand messages to this AMQP queue which it had received via WebSockets on another channel
  servers:
    - $ref: "#/components/servers/rabbitmqBrokerInProd"
    - $ref: "#/components/servers/rabbitmqBrokerInStaging"
  subscribe:
    message:
      $ref: "#/components/messages/WebUICommand"
  bindings:
    amqp:
      is: queue
...

components:
  servers:
    rabbitmqBrokerInProd:
      url: staging.gigantic-server.com
      description: Staging server
      protocol: amqp
      protocolVersion: 0.9.1
   rabbitmqBrokerInStaging
      url: api.gigantic-server.com
      description: Production server
      protocol: amqp
      protocolVersion: 0.9.1
    wsDev:
      ...
    wsProd:
      ...
```

Please review https://github.com/asyncapi/spec/pull/531/files and approve.

File: spec/asyncapi.md [hunk 1]
```diff
@@ -640,6 +656,40 @@ subscribe:
 ```
 
 
+Using eplicit references to the servers on which the channel is available:
```
Comment (fmvilas) [inline spec/asyncapi.md:659]:
```suggestion
Using explicit references to the servers on which the channel is available:
```

File: spec/asyncapi.md [hunk 2]
```diff
@@ -1322,6 +1372,7 @@ All objects defined within the components object will have no effect on the API
 
 Field Name | Type | Description
 ---|:---|---
+<a name="componentsServers"></a> servers | Map[`string`, [Server Object](#serverObject) \| [Reference Object](#referenceObject)] | An object to hold reusable [Server Objects](#serverObject).
```
Comment (fmvilas) [inline spec/asyncapi.md:1375]: Do we really need this? Why can't we point directly to `#/servers/{name-of-server}`?

File: spec/asyncapi.md [hunk 3]
```diff
@@ -640,6 +656,40 @@ subscribe:
...
+    }
+  }
+}
+```
+
+```yaml
+description: This application publishes WebUICommand messages to this AMQP queue which it had received via WebSockets on another channel
+servers:
+  - $ref: "#/components/servers/rabbitmqBrokerInProd"
+  - $ref: "#/components/servers/rabbitmqBrokerInStaging"
```
Comment (fmvilas) [inline spec/asyncapi.md:685]: Why not like this:

```suggestion
  - $ref: "#/servers/amqpBrokerInProd"
  - $ref: "#/servers/amqpBrokerInStaging"
```

**Review by fmvilas — CHANGES_REQUESTED** [review]
Left some comments, Gerald. I think we can simplify this even a bit more.

**Comment by fmvilas** [comment]
Also, related to your comment in #513, I think it's too early for this proposal to get into the next release. It's still a Strawman (RFC 0) proposal and I don't think we're giving enough time for people to review and implement support in the parsers.

File: spec/asyncapi.md (hunk 2 above)
Comment (GeraldLoeffler) [reply spec/asyncapi.md:1375]: the only other instances i found in the spec where a field points to another field where when they pointed to component objects: this is the only reason why i've modelled it this way. If you say we can point directly to `#/servers/{name-of-server}` and get the same kind of static validation of the reference then i'm more than happy to make that change. Let me know @fmvilas .

**Comment by GeraldLoeffler** [comment]
> Also, related to your comment in #513, I think it's too early for this proposal to get into the next release. It's still a Strawman (RFC 0) proposal and I don't think we're giving enough time for people to review and implement support in the parsers.

that's very unfortunate, but there's nothing i can do about it other than try to shepherd this through the required process...

File: spec/asyncapi.md (hunk 3 above)
Comment (GeraldLoeffler) [reply spec/asyncapi.md:685]: i understand this is your proposal to linking directly to `#/servers/{name-of-server}`: but `{name-of-server}` (eg, `amqpBrokerInProd`) in this case would be name of a server in the map of server objects (https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#servers-object): if you say this works just as well as pointing to a components object then i'm OK with that.

**Comment by GeraldLoeffler** [comment]
i've now implemented the further simplification suggested by @fmvilas , which looks like this:
```
servers:
  rabbitmqBrokerInProd:
    url: staging.gigantic-server.com
    description: Staging server
    protocol: amqp
    protocolVersion: 0.9.1
 rabbitmqBrokerInStaging
    url: api.gigantic-server.com
    description: Production server
    protocol: amqp
    protocolVersion: 0.9.1
  wsDev:
    ...
  wsProd:
    ...

webUICommandsQueue:
  description: This application publishes to this AMQP queue WebUICommand messages which it had received via WebSockets on another channel
  servers:
    - $ref: "#/servers/rabbitmqBrokerInProd"
    - $ref: "#/servers/rabbitmqBrokerInStaging"
  subscribe:
    message:
      $ref: "#/components/messages/WebUICommand"
  bindings:
    amqp:
      is: queue
...
```

I think this is now the simplest possible design that still allows AsyncAPI document authors to assign channels to a sub-set of the defined servers.

@derberg , @fmvilas : please review and merge

**Comment by GeraldLoeffler** [comment]
please dear gods of GitHub, unstale this PR! - I still think this can/should be merged ASAP, because it is fully backwards-compatible: i'm hoping for the September release of AsyncAPI...

**Comment by derberg** [comment]
I think this proposal can reach `Stage 1: Proposal`
- it is already opened for 3 months
- there was a good discussion
- we shared it with the community on all possible channels, including https://www.asyncapi.com/blog/may-2021-at-asyncapi#assigning-channels-to-servers
- I assume @GeraldLoeffler is obviously a champion here that will drive it further, including changes in the JS Parser and JSON Schema
- all are clearly explained, and the use case is pretty obvious. In slack we heard many people asking for such support on channel level

From the definition of Stage 1 all requirements are met imho. Opinions @fmvilas ?

I still have some concerns (non-bloking promotion to Stage 1) with approach that we use refs to Servers Object here @fmvilas. Spec says that `components` are there for reuse purposes, so this can cause confusion that servers are reused without components. Unless, unless this way we want to make sure the user provides `Servers Object`. We set a kind of requirement on the user that to use a specific server on a channel level, all servers need to be listed under `Servers Object`. I don't see why it is needed though. Isn't it an unnecessary restriction? 🤔 Somehow I have a feeling that in case we add Servers on Components, usage of ref definitely makes sense, but using ref on Servers directly sounds weird 🤷🏼 I kinda immediately think it should be just a list of names (or label like in case of initial request)

**Comment by magicmatatjahu** [comment]
I wasn't involved with this proposal from the beginning, but as I read the discussion two things wonder me: 
- what in the case if someone add reference to the server, but from the other AsyncAPI doc, like `$ref: ./asyncapi-two.yaml/servers/production`. In the current proposal it's possible, but there should be only possibility to reference to the servers from current AsyncAPI document, so as @derberg you wrote:

  > The advantage is that you have validation out of the box, parsing fails if you want to parse outdated reference...
  
  We must also implement custom parsing for this case - check if references server is included in the `servers` section.

- why not using the normal server names, without any labels/references, similar like in https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#security-requirement-object?

  ```yaml
  servers:
    rabbitmqBrokerInProd:
      url: staging.gigantic-server.com
      description: Staging server
      protocol: amqp
      protocolVersion: 0.9.1
   rabbitmqBrokerInStaging
      url: api.gigantic-server.com
      description: Production server
      protocol: amqp
      protocolVersion: 0.9.1
    wsDev:
      ...
    wsProd:
      ...
      
  webUICommandsQueue:
    description: This application publishes to this AMQP queue WebUICommand messages which it had received via WebSockets on another channel
    servers: ['rabbitmqBrokerInProd', 'rabbitmqBrokerInStaging']
    subscribe:
      message:
        $ref: "#/components/messages/WebUICommand"
    bindings:
      amqp:
        is: queue
  ```
  
Maybe there are answers to these questions in the attached issues or in this discussion, but I have not found them.

**Comment by fmvilas** [comment]
Yeah, I think this can be considered a Stage 1 Proposal.

I was actually thinking that it may be better to use server names instead of `$ref`. The problem with $ref is that we lose the name of the server when it's dereferenced.

For instance:

```yaml
asyncapi: 2.1.0
servers:
  serverA:
    url: ws://server-a.com
  serverB:
    url: ws://server-b.com

channels:
  mychannel:
    servers:
      - $ref: '#/servers/serverA'
```

becomes

```yaml
asyncapi: 2.1.0
servers:
  serverA:
    url: ws://server-a.com
  serverB:
    url: ws://server-b.com

channels:
  mychannel:
    servers:
      - url: ws://server-a.com
```

As you can see, in the second example, we lose the `serverA` identifier of the server.

I realized about this while working on Glee. In Glee, you can use a `x-servers` extension (not documented yet) to define which servers you want to associate with the channel. It's user-configurable too so having the `serverA` identifier is necessary.

So, should we instead use plain strings to refer to the servers? Example:

```yaml
asyncapi: 2.1.0
servers:
  serverA:
    url: ws://server-a.com
  serverB:
    url: ws://server-b.com

channels:
  mychannel:
    servers:
      - serverA
```

**Comment by magicmatatjahu** [comment]
@fmvilas Great minds think the same 😄

**Comment by GeraldLoeffler** [comment]
> So, should we instead use plain strings to refer to the servers?

string-references would not be validated at design-time by a YAML editor like `$ref`s would, correct?

**Comment by fmvilas** [comment]
> string-references would not be validated at design-time by a YAML editor like `$refs` would, correct?

`$ref` is not part of YAML specification so I guess a YAML editor would behave the same either with `$ref` and plain string references.

**Comment by magicmatatjahu** [comment]
@GeraldLoeffler won't be, but with using `$ref` there are problems described by me and Fran in previous comments:

- how to check if given server is in current AsyncAPI doc?
- what in case when someone define plain server in the `channel/servers` field, without ref? 

So at the end we still need custom validation.

**Comment by fmvilas** [comment]
I'm confused. How is a YAML editor helping you with `$ref`s? I'm assuming you're referring to specific OpenAPI/AsyncAPI/JSON Schema editors, right? If that's the case, these editors would have to implement support for validation at design time. Just the same way they implemented support for `$ref`.

**Comment by GeraldLoeffler** [comment]
> I'm confused. How is a YAML editor helping you with `$ref`s? I'm assuming you're referring to specific OpenAPI/AsyncAPI/JSON Schema editors, right? If that's the case, these editors would have to implement support for validation at design time. Just the same way they implemented support for `$ref`.

VSCode, for example, with the Red Hat YAML extension, understands YAML and JSON Reference and JSON Pointer, so it knows if a `$ref` is correct, and if it is it shows it as a hyperlink that you can command-click to follow:
yaml-with-ref

**Comment by derberg** [comment]
+1 for server names instead of $refs because of dereferencing. I don't really get why `how to check if given server is in current AsyncAPI doc?` would be a problem. 

and yes, there will have to be a custom validation added to the parser to check if the name added in the channel server matches the server from servers. We already perform such custom validation for server security, if it matches security from components.

the lack of support in YAML editor would be solved once we invest more time in https://github.com/asyncapi/vs-asyncapi-preview. Anyway, not related to this proposal imho.

**Comment by magicmatatjahu** [comment]
> I don't really get why `how to check if given server is in current AsyncAPI doc?` would be a problem.

If we will use the name of server, it won't be a problem, but it will be with references. After dereferencing you will end with object of the server (without name) and additionally as Fran wrote, you will be able  write the plain server in the `servers` field of channel, so there is only one possibility to check that given server is defined in current doc... by deeply checking the entire server shape (or checking the url). In the string case, you only check the array of server names.

**Comment by GeraldLoeffler** [comment]
changed this proposal to refer to servers by name rather than by ref (`$ref$).

**Review by fmvilas — CHANGES_REQUESTED** [review]

**Comment by GeraldLoeffler** [comment]
@fmvilas : made change requested by you and newly fetched from upstream (`asyncapi:master`)

**Comment by GeraldLoeffler** [comment]
@fmvilas : this PR was still in state "changes requested" (by you), but **i have already addressed all requested changes**: i've therefore **re-requested your review**. In other words: i'm done with this and i think it addresses all requests and comments by you and @derberg and @magicmatatjahu .

File: spec/asyncapi.md [hunk 4]
```diff
@@ -562,7 +562,8 @@ Describes the operations available on a single channel.
 Field Name | Type | Description
 ---|:---:|---
 <a name="channelItemObjectRef"></a>$ref | `string` | Allows for an external definition of this channel item. The referenced structure MUST be in the format of a [Channel Item Object](#channelItemObject). If there are conflicts between the referenced definition and this Channel Item's definition, the behavior is *undefined*.
-<a name="channelItemObjectDescription"></a>description | `string` | An optional description of this channel item. [CommonMark syntax](https://spec.commonmark.org/) can be used for rich text representation.
+<a name="channelItemObjectDescription"></a>description | `string` | An optional description of this channel item. [CommonMark syntax](http://spec.commonmark.org/) can be used for rich text representation.
```
Comment (fmvilas) [inline spec/asyncapi.md:565]:
```suggestion
<a name="channelItemObjectDescription"></a>description | `string` | An optional description of this channel item. [CommonMark syntax](https://spec.commonmark.org/) can be used for rich text representation.
```

**Review by fmvilas — CHANGES_REQUESTED** [review]
Minor thing. Otherwise 👍

**Comment by GeraldLoeffler** [comment]
> Minor thing. Otherwise 👍

requested fix accepted!

**Review by magicmatatjahu — APPROVED** [review]

**Comment by derberg** [comment]
@GeraldLoeffler all good from my side here but PR is against `master`, so please change base branch to release branch and rebase

**Comment by derberg** [comment]
as for the other PRs, JSON Schema merged and RC produced. We are missing full implementation in parser -> https://github.com/asyncapi/parser-js/pull/364. So far we are at Stage 1 here

**Comment by GeraldLoeffler** [comment]
@derberg : rebased to release branch - done

File: spec/asyncapi.md [hunk 5]
```diff
@@ -563,6 +563,7 @@ Field Name | Type | Description
 ---|:---:|---
 <a name="channelItemObjectRef"></a>$ref | `string` | Allows for an external definition of this channel item. The referenced structure MUST be in the format of a [Channel Item Object](#channelItemObject). If there are conflicts between the referenced definition and this Channel Item's definition, the behavior is *undefined*.
 <a name="channelItemObjectDescription"></a>description | `string` | An optional description of this channel item. [CommonMark syntax](https://spec.commonmark.org/) can be used for rich text representation.
+<a name="channelItemObjectServers"></a>servers | [[Server Object](#serverObject)] | The servers on which this channel is available, specified as an optional unordered list of names (string keys) of [Server Objects](#serverObject) defined in the [Servers Object](#serversObject) (a map). If `servers` is absent or empty then this channel must be available on all servers defined in the [Servers Object](#serversObject).
```
Comment (derberg) [inline spec/asyncapi.md:566]:
```suggestion
<a name="channelItemObjectServers"></a>servers | [`string`] | The servers on which this channel is available, specified as an optional unordered list of names (string keys) of [Server Objects](#serverObject) defined in the [Servers Object](#serversObject) (a map). If `servers` is absent or empty then this channel must be available on all servers defined in the [Servers Object](#serversObject).
```

I think this type is leftover from the initial proposal where we advised that `servers` is an array of objects or references, while now it is actually an array of strings

**Review by derberg — COMMENTED** [review]
A reference implementation in JSON Schema and Parser are already released in individual release candidates. I changed this proposal to Stage 3.

There is one error IMHO here in this PR, please have a look at the individual comment to markdown file cc @fmvilas @magicmatatjahu

File: spec/asyncapi.md (hunk 5 above)
Comment (magicmatatjahu) [reply spec/asyncapi.md:566]: I agree! In addition, the PR description would have to be changed, as it still points to the old idea related to labels.

**Comment by derberg** [comment]
@GeraldLoeffler fyi we cannot make fixes on our own as you did PR from some kind of org-level fork and we do not have rights as maintainer to push to this PR.

**Comment by GeraldLoeffler** [comment]
> @GeraldLoeffler fyi we cannot make fixes on our own as you did PR from some kind of org-level fork and we do not have rights as maintainer to push to this PR.

i accepted your suggestion and gave you write access to my repo.

**Comment by derberg** [comment]
@GeraldLoeffler perfect, the only part that is missing is updated description that reflects what is actually implemented

**Comment by GeraldLoeffler** [comment]
> @GeraldLoeffler perfect, the only part that is missing is updated description that reflects what is actually implemented

done

**Review by derberg — APPROVED** [review]
👏🏼

**Comment by fmvilas** [comment]
Super happy to see thins merged 👏

**Pull request merged**